src/checkers/inference/util/JaifBuilder.java

```diff
@@ -240,7 +240,12 @@ private void buildClassEntries() {
                     // TODO: this is not a feature but a workaround of a bug:
                     // We should create a non-empty correct ASTPath for constructor
                     if (astRecord.astPath.equals(ASTPath.empty())) {
-                        continue;
+                        if (astRecord.methodName != null && astRecord.methodName.contains("<init>")) {
```

Comment:
Is this change needed or is the issue already solved by the change to `getConstructorRecord`?
Is the difference that `getConstructorRecord` is only for no-arg constructors and this code also handles constructors with parameters?
Wouldn't a better fix be to correctly construct the right ASTRecord, instead of patching up the result when outputting it?

Reply:
The only place in CFI that calls `ASTPathUtil#getConstructorRecord()` is: https://github.com/opprop/checker-framework-inference/blob/bd084ced09b5b4fa2f78b5538400e9cebc53ac91/src/checkers/inference/VariableAnnotator.java#L1412
For me, it seems like this line executes only for non-top level constructors, so I doubt `ASTPathUtil#getConstructorRecord()` doesn't fix the constructor return type problem without this line of fix.
The fix here applies to all constructors(with or without parameters), and patches correct `ASTRecord` before writing to jaif file.
CFI doesn't have any logic of building `ASTRecord` for source trees: it only calls method `ASTInde#indexOf(CompilationUnitTree)` to get the caches from annotation-tools.
The problem comes from this line in annotation-tools: https://github.com/opprop/annotation-tools/blob/master/scene-lib/src/scenelib/annotations/io/ASTIndex.java#L381
`MethodTree#getReturnType()` for constructor is null, so its `ASTRecord` is not extended with "Method, TYPE" for return.
We should either fix it here in annotation-tools or at:
https://github.com/opprop/checker-framework-inference/blob/bd084ced09b5b4fa2f78b5538400e9cebc53ac91/src/checkers/inference/VariableAnnotator.java#L189
If we detect empty `ASTPath` in `record` and the argument tree `tree` is constructor, we create valid "Method TYPE" `ASTPath` to `record`. Which do you think is better? @wmdietl